src/core/or/policies.c

```diff
@@ -495,8 +500,39 @@ fascist_firewall_prefer_ipv6_impl(const or_options_t *options)
 MOCK_IMPL(int,
 fascist_firewall_rand_prefer_ipv6_addr, (void))
 {
-  /* TODO: Check for failures, and infer our preference based on this. */
-  return crypto_rand_int(2);
+  or_state_t *state = get_or_state();
+  time_t current_time = time(NULL);
+
+  /* Clear IPv4 and IPv6 failure entries if they are more than a week old. */
+  if (state->IPv4AutoFailTime &&
+      state->IPv4AutoFailTime <= current_time - IP_FAIL_CLEAR_TIME) {
+    state->IPv4AutoFail = 0;
+    state->IPv4AutoFailTime = 0;
+  }
+  if (state->IPv6AutoFailTime &&
+      state->IPv6AutoFailTime <= current_time - IP_FAIL_CLEAR_TIME) {
+    state->IPv6AutoFail = 0;
+    state->IPv6AutoFailTime = 0;
+  }
+
+  if (state->IPv4AutoFail && !state->IPv6AutoFail) {
```

Comment:
We don't want to turn off an IP family just because our routing table doesn't have an entry for one IP address. Maybe we should track the number of failed connections due to no route to host, and give them a higher weight.